colossalai/zero/sharded_optim/sharded_optim_v2.py

```diff
@@ -358,3 +362,28 @@ def _copy_master_param_to_param_fp16(self):
                 p.data = p.colo_attr.sharded_data_tensor.payload
                 self.master_params[p].trans_state(TensorState.HOLD)
                 p.colo_attr.saved_grad.set_null()
+
+    def _set_optim_states_before_step(self):
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                # We don't need to set
+                # self.optim.state[p][k] = self._state[p][k].payload
+                # since payload is moved over devices inline
+                # self.optim.state[p][k]'s tensor id won't change
+                # the only change is its data pointer
+                for optim_state in self._state[p].values():
+                    optim_state.trans_state(TensorState.COMPUTE)
```

Comment:
In this way, you set all OS tensor to compute?
You should set a tensor to COMPUTE the exact time it is about to be computed.

Reply:
We are unable to do this unless we use a custom optimizer.

Reply:
That is to say, users cannot use torch.optim.SGD, torch.optim.Adam anymore.

Reply:
I am afraid so. If you use the stateful tensor, then you have to use our stateful optimizer.

Reply:
I discuss with @1SAA . We both think that OS is not suitable for dynamic memory layout. Static layout like now maybe more  efficient.

Reply:
Move OS dynamically meybe not reduce computation time, however, it increases data movement time. Besides, it won't save momery usage.

Reply:
And we have to make sure OS, fp32 weight and gradient are on the same device for each parameter. That is to say, our eviction algorithm has to identify those stateful tensors are "linked" to which parameter. Although they are all "tensors", but they are not "independent" tensors.

Reply:
I think dynamic memory layout is suitable for fp16 params. OS, fp32 weight and gradient can be moved once and stay on the device.

Reply:
OK, we make param fp16 stateful tensor first.